src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -417,19 +400,161 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
+    let mut candidates: Vec<_> = implicit_impls
+        .flat_map(|impl_outer| {
+            match impl_outer.inner {
+                ImplItem(impl_) => {
+                    debug!("considering auto or blanket impl for trait {:?}", impl_.trait_);
+                    // Give precedence to methods that were overridden
+                    if !impl_.provided_trait_methods.contains(&*item_name.as_str()) {
+                        let mut items = impl_.items.into_iter().filter_map(|assoc| {
+                            if assoc.name.as_deref() != Some(&*item_name.as_str()) {
+                                return None;
+                            }
+                            let kind = assoc
+                                .inner
+                                .as_assoc_kind()
+                                .expect("inner items for a trait should be associated items");
+                            if kind.namespace() != ns {
+                                return None;
+                            }
+
+                            trace!("considering associated item {:?}", assoc.inner);
+                            // We have a slight issue: normal methods come from `clean` types,
+                            // but provided methods come directly from `tcx`.
+                            // Fortunately, we don't need the whole method, we just need to know
+                            // what kind of associated item it is.
+                            Some((assoc.def_id, kind))
+                        });
+                        let assoc = items.next();
+                        debug_assert_eq!(items.count(), 0);
+                        assoc
+                    } else {
+                        // These are provided methods or default types:
+                        // ```
+                        // trait T {
+                        //   type A = usize;
+                        //   fn has_default() -> A { 0 }
+                        // }
+                        // ```
+                        let trait_ = impl_.trait_.unwrap().def_id().unwrap();
+                        cx.tcx
+                            .associated_items(trait_)
+                            .find_by_name_and_namespace(
+                                cx.tcx,
+                                Ident::with_dummy_span(item_name),
+                                ns,
+                                trait_,
+                            )
+                            .map(|assoc| (assoc.def_id, assoc.kind))
+                    }
+                }
+                _ => panic!("get_impls returned something that wasn't an impl"),
+            }
+        })
+        .collect();
+
+    // Next consider explicit impls: `impl MyTrait for MyType`
+    // Give precedence to inherent impls.
+    if candidates.is_empty() {
+        // FIXME: As an optimization, this could only search traits in scope.
```

Comment:
Yeah, to be clear we have to make the decision on this now. The reason I _wanted_ it was as an optimization, but we cannot make it later once this lands.

Reply:
I'm fine with restricting it to traits in scope, but I don't know _how_. I asked about this a while ago and never got a response: https://github.com/rust-lang/rust/pull/74489#issuecomment-666029654

Reply:
Maybe it's related to cross-crate links somehow and it can only tell the traits in scope for the current crate? I don't want to have different behavior in re-exported links than local links.

Reply:
The other thing I could do is go through `resolve_str_path_error` like the rest of the links, which behaves the same cross-crate. That loses all the efficiency benefits though: I have to first go through all traits and see if they _could_ be a match, then find whether they can be resolved in scope.

Reply:
^ This doesn't work because the trait could be renamed (or unnamed: `use std::io::Write as _;`).

@petrochenkov do you know a way to find whether a trait is in scope across crates? I know `in_scope_traits` works for the current crate, but I don't know how to do it for others. This is for a module, rustdoc doesn't need the information for individual expressions.

Reply:
Yay I think I found a way to do this 🎉 [`get_traits_containing_item`](https://github.com/rust-lang/rust/blob/master/src/librustc_resolve/late.rs#L2310-L2412)

Reply:
I changed the behavior so this now only looks at traits in scope (efficiently, without having to look at every possibly trait). This is what needed the changes to `rustc_resolve`.